cryptol/readme.md

```diff
@@ -0,0 +1,188 @@
+        • out == pad
+            arising from
+            matching types
+```
+
+Since we use the types `pad`, `in`, and `out` to specify lengths of strings and in functions requiring finite values, we must mark them finite with the `fin` constraint.
+Moreover, because we used `pad - in` as the length of the prefix, it must be non-negative; hence `pad` must be greater than or equal to `in`.
+Finally, the interpreter deduced that the length `out` must be the same as `pad`, which we can check for ourselves from our desired properties of left pad: we need `out` to be the maximum of `pad` and `in`, but `pad >= in` means that `pad` _is_ the maximum of `pad` and `in`.
+
+For ease of reuse, let's factor our the length constraints, and then use them in our definition of `leftpad`.
+
+```cryptol
+type constraint lpCons pad in out = (fin pad, fin in, fin out, pad >= in, out == pad)
+
+leftpad : {pad, in, out} (lpCons pad in out) => Char -> String in -> String out
+leftpad c s = repeat `{pad - in} c # s
+```
+
+With this, asking `cryptol` to load the file completes without complaint:
+
+```
+$ cryptol
+┏━╸┏━┓╻ ╻┏━┓╺┳╸┏━┓╻
+┃  ┣┳┛┗┳┛┣━┛ ┃ ┃ ┃┃
+┗━╸╹┗╸ ╹ ╹   ╹ ┗━┛┗━╸
+version 2.13.0
+https://cryptol.net  :? for help
+
+Loading module Cryptol
+Cryptol> :l leftpad.cry
+Loading module Cryptol
+Loading module leftpad
+```
+
+Now for the properties!
+I'll include the first one here for our discussion; what follows will also apply to the other two properties as well.
+
+Here's the first property in all its glory, reusing our constraint from above:
+
+```cryptol
+leftpadLength : {pad, in, out} (lpCons pad in out) => Char -> String in -> Bit
+property leftpadLength c s = length (leftpad `{pad, in, out} c s) == `out
+```
+
+Recalling the constraints `lpCons`, this says that the length of our output string is `out` A.K.A. `pad` A.K.A. `max(pad, in)`.
+
+Trying to `:prove` this property, however, the interpreter complains:
+
+```
+leftpad> :prove leftpadLength
+Not a monomorphic type:
+{pad, in, out}
+  (fin pad, fin in, fin out, pad >= in, out == pad) =>
+  [8] -> [in][8] -> Bit
+(Total Elapsed Time: 0.000s)
+```
+
+Because the type `{pad, in, out}` is not monomorphic (i.e. instantiated with specific numbers in this case), `cryptol` refuses to try and prove `leftpadLength`.
+Quoth [Programming Cryptol](https://cryptol.net/files/ProgrammingCryptol.pdf),
+
+> we cannot directly prove polymorphic properties as they may hold for certain monomorphic instances while not for others. In this cases, we must tell Cryptol what particular monomorphic instance of the property we would like it to prove.
+
+I suppose this makes sense given Cryptol's intended use.
+Though we may wish to state polymorphic constraints with type variables like `pad`, `in`, and `out`, Cryptol's ultimate goal is to construct rock-solid specifications of cryptographic algorithms; in "real world implementations," one would fill in these type variables for any specific implementation.
```

Comment:
Maybe a quick mention of why monomorphic instances are so important- it's side channel attacks, I think?

Reply:
I suppose there could be side-channel leakage—maybe if an operation depended on the value or bits of a specific type variable?—but I don't think this was the primary concern. I think the main issue is that the ability to express _arbitrary_ arithmetic at the type level leads to undecidability; per [Programming Cryptol](https://cryptol.net/files/ProgrammingCryptol.pdf)

> Type inference in the presence of arithmetic predicates is an undecidable problem [6]. This implies that there
is no algorithm to decide whether a given type is inhabited, amongst other things. In practical terms, we might end up writing programs with arbitrarily complicated predicates (e.g., this “type contains all solutions to Fermat’s last equation” or “this type contains all primes between two large numbers”) without Cryptol being able to simplify them, or notice that there is no instantiation that will ever work.